File: backend/app/models/fiado.py
from sqlalchemy import Column, String, Numeric, DateTime, Boolean, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid
from app.database import Base


class Cliente(Base):
    __tablename__ = "clientes"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    comercio_id = Column(UUID(as_uuid=True), ForeignKey("comercios.id"), nullable=False)
    nome = Column(String(255), nullable=False)
    telefone = Column(String(20))
    criado_em = Column(DateTime, default=datetime.utcnow)

    fiados = relationship("Fiado", back_populates="cliente")


class Fiado(Base):
    __tablename__ = "fiados"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    cliente_id = Column(UUID(as_uuid=True), ForeignKey("clientes.id"), nullable=False)
    comercio_id = Column(UUID(as_uuid=True), ForeignKey("comercios.id"), nullable=False)
    descricao = Column(String(255))
    valor = Column(Numeric(10, 2), nullable=False)
    pago = Column(Boolean, default=False)
    data_venda = Column(DateTime, default=datetime.utcnow)
    data_pagamento = Column(DateTime, nullable=True)

    cliente = relationship("Cliente", back_populates="fiados")

File: worker/src/routes/fiado.ts
import { Hono } from "hono";
import { createDb, clientes, fiados } from "../db";
import { eq, and, sum } from "drizzle-orm";
import type { Bindings } from "../index";
import { randomUUID } from "crypto";

export const fiadoRoutes = new Hono<{ Bindings: Bindings }>();

fiadoRoutes.get("/clientes", async (c) => {
  const db = createDb(c.env.DB);
  const comercioId = c.req.query("comercio_id") ?? "";
  const lista = await db
    .select()
    .from(clientes)
    .where(eq(clientes.comercioId, comercioId));
  return c.json(lista);
});

fiadoRoutes.post("/clientes", async (c) => {
  const db = createDb(c.env.DB);
  const body = await c.req.json();
  const novo = {
    id: randomUUID(),
    comercioId: body.comercio_id,
    nome: body.nome,
    telefone: body.telefone ?? null,
    criadoEm: Date.now(),
  };
  await db.insert(clientes).values(novo);
  return c.json(novo, 201);
});

fiadoRoutes.get("/clientes/:clienteId/fiados", async (c) => {
  const db = createDb(c.env.DB);
  const lista = await db
    .select()
    .from(fiados)
    .where(eq(fiados.clienteId, c.req.param("clienteId")));
  return c.json(lista);
});

fiadoRoutes.post("/", async (c) => {
  const db = createDb(c.env.DB);
  const body = await c.req.json();
  const novo = {
    id: randomUUID(),
    clienteId: body.cliente_id,
    comercioId: body.comercio_id,
    descricao: body.descricao ?? null,
    valor: body.valor,
    pago: false,
    dataVenda: Date.now(),
    dataPagamento: null,
  };
  await db.insert(fiados).values(novo);
  return c.json(novo, 201);
});

fiadoRoutes.patch("/:fiadoId/pagar", async (c) => {
  const db = createDb(c.env.DB);
  await db
    .update(fiados)
    .set({ pago: true, dataPagamento: Date.now() })
    .where(eq(fiados.id, c.req.param("fiadoId")));
  return c.json({ mensagem: "Fiado marcado como pago" });
});
